docs/study-def-tricks.md

```diff
@@ -321,63 +321,68 @@ study = StudyDefinition(
+    default_expectations={
+        "date": {"earliest": "1900-01-01", "latest": "today"},
+        "rate": "uniform",
+        "incidence": 0.5,
+    },
+
+    # We extract ethnicity once for all patients, to ensure we don't exclude
+    # any patients.
+    population=patients.all(),
+
+    ethnicity=patients.with_these_clinical_events(
+        ethnicity_codes,  # the codelist
+        returning="category",
+        find_last_match_in_period=True,  # most recent code
+        return_expectations={
+            "category": {"ratios": {"1": 0.2, "2": 0.2, "3": 0.2, "4": 0.2, "5": 0.2}},
+            "incidence": 0.75,
+        },
+    ),
+)
 ```
 
-You can then join the outputs of this study definition using a function like this:
+The following [scripted action](actions-scripts.md) (`analysis/join_ethnicity.py`) demonstrates how we would join each index date's extract to the ethnicity extract:
 
-`join_ethnicity.py`
-```py
-    from pathlib import Path
+```python
+import pathlib
 
-    def join_ethnicity(directory: str) -> None:
-    """Finds 'input_ethnicity.feather' in directory and combines with each input file."""
-    
-        dirpath = Path(directory)
-        filelist = dirpath.iterdir()
-
-        # get ethnicity input file
-        ethnicity_df = pd.read_feather(dirpath / 'input_ethnicity.feather')
-        
-        # map ethnicity using the patient_id
-        ethnicity_dict = dict(zip(ethnicity_df['patient_id'], ethnicity_df['ethnicity']))
-        
-        # loop over files in directory and overwrite with a file containing ethnicity
-        for file in filelist:
-            if match_input_files(file.name):
-                df = pd.read_feather(dirpath / file.name)
-            
-                df['ethnicity'] = df['patient_id'].map(ethnicity_dict)
-                df.to_feather(dirpath / file.name) # overwrites existing dataframe with ones which include ethnicity
-
-    join_ethnicity('output')
+import pandas
+
+
+path_to_output = pathlib.Path("output")
+
+input_ethnicity = pandas.read_feather(path_to_output / "input_ethnicity.feather")
+
+for path in path_to_output.iterdir():
+    if not path.name.startswith("input") or path.name == "input_ethnicity.feather":
+        continue
+
+    input_other = pandas.read_feather(path)
+    input_other = input_other.merge(input_ethnicity, how="left", on="patient_id")
+
+    # If the old path was "output/input_2021-01-01.feather", then the new
+    # path will be "output/input_2021-01-01_merged.feather".
+    new_path = path_to_output / f"{input_other.stem}_merged{input_other.suffix}"
+    input_other.to_feather(new_path)
 ```
 
-You can find an example of this in [this repo](https://github.com/opensafely/sro-template) (using Python) or in [this repo](https://github.com/opensafely/covid_mortality_over_time) (using R).
+You can find a Python example in [this repo](https://github.com/opensafely/sro-template) and an R example in [this repo](https://github.com/opensafely/sro-template).
```

Comment:
question: is the link to the R example correct here? I checked the linked repository but can't find any R code there (unless I'm missing something). Or is it supposed to be the [original link](https://github.com/opensafely/covid_mortality_over_time)?

Reply:
suggestion: Perhaps consider make the "this repo" link text more useful, especially as it's duplicated in the same sentence. For example:

```suggestion
### Research examples

Refer to existing research repositories using this extraction approach:

* [Python code](https://github.com/opensafely/sro-template)
* [R code](https://github.com/opensafely/covid_mortality_over_time).
```

The above change :arrow_up: assumes that the original link to the R code is correct, and the change to `sro-template` is a mistake.